File: src/test/java/de/fraunhofer/aisec/codyze/crymlin/AnalysisServerBotanTest.java

package de.fraunhofer.aisec.codyze.crymlin;

import de.fraunhofer.aisec.codyze.analysis.passes.EdgeCachePass;
import de.fraunhofer.aisec.codyze.analysis.passes.IdentifierPass;
import de.fraunhofer.aisec.codyze.analysis.AnalysisServer;
import de.fraunhofer.aisec.codyze.analysis.AnalysisContext;
import de.fraunhofer.aisec.codyze.analysis.ServerConfiguration;
import de.fraunhofer.aisec.cpg.TranslationConfiguration;
import de.fraunhofer.aisec.cpg.TranslationManager;
import de.fraunhofer.aisec.codyze.markmodel.MEntity;
import de.fraunhofer.aisec.codyze.markmodel.MRule;
import de.fraunhofer.aisec.codyze.markmodel.Mark;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.io.File;
import java.net.URL;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

import static org.junit.jupiter.api.Assertions.*;

class AnalysisServerBotanTest {

	private static AnalysisServer server;
	private static AnalysisContext result;

	@BeforeAll
	public static void startup() throws Exception {
		ClassLoader classLoader = AnalysisServerBotanTest.class.getClassLoader();

		URL resource = classLoader.getResource("mark_cpp/symm_block_cipher.cpp");
		assertNotNull(resource);
		File cppFile = new File(resource.getFile());
		assertNotNull(cppFile);

		resource = classLoader.getResource("mark/PoC_MS1/Botan_AutoSeededRNG.mark");
		assertNotNull(resource);
		File markPoC1 = new File(resource.getFile());
		assertNotNull(markPoC1);
		String markModelFiles = markPoC1.getParent();

		// Start an analysis server
		server = AnalysisServer.builder()
				.config(
					ServerConfiguration.builder().launchConsole(false).launchLsp(false).markFiles(markModelFiles).build())
				.build();
		server.start();

		// Start the analysis (BOTAN Symmetric Example by Oliver)
		TranslationManager translationManager = newAnalysisRun(cppFile);
		CompletableFuture<AnalysisContext> analyze = server.analyze(translationManager);
		try {
			result = analyze.get(5, TimeUnit.MINUTES);
		}
		catch (TimeoutException t) {
			analyze.cancel(true);
			throw t;
		}
	}

	@AfterAll
	public static void teardown() {
		// Stop the analysis server
		server.stop();
	}

	@Test
	void markModelTest() {
		Mark markModel = server.getMarkModel();
		assertNotNull(markModel);
		List<MRule> rules = markModel.getRules();
		assertEquals(7, rules.size());

		Collection<MEntity> ents = markModel.getEntities();
		assertEquals(9, ents.size());
	}

	@Test
	void markEvaluationTest() {
		AnalysisContext ctx = AnalysisServerBotanTest.result;
		assertNotNull(ctx);
		List<String> findings = new ArrayList<>();
		assertNotNull(ctx.getFindings());
		ctx.getFindings().forEach(x -> findings.add(x.toString()));

		System.out.println("Findings");
		for (String finding : findings) {
			System.out.println(finding);
		}
	}

	/**
	 * Helper method for initializing an Analysis Run.
	 *
	 * @param sourceLocations
	 * @return
	 */
	private static TranslationManager newAnalysisRun(File... sourceLocations) {
		return TranslationManager.builder()
				.config(
					TranslationConfiguration.builder()
							.debugParser(true)
							.failOnError(false)
							.defaultPasses()
							.registerPass(new IdentifierPass())
							.registerPass(new EdgeCachePass())
							.defaultLanguages()
							.sourceLocations(sourceLocations)
							.build())
				.build();
	}
}

File: src/test/java/de/fraunhofer/aisec/codyze/crymlin/AnalysisServerBotanTest.kt
package de.fraunhofer.aisec.codyze.crymlin

import de.fraunhofer.aisec.codyze.analysis.AnalysisContext
import de.fraunhofer.aisec.codyze.analysis.AnalysisServer
import de.fraunhofer.aisec.codyze.analysis.Finding
import de.fraunhofer.aisec.codyze.analysis.ServerConfiguration
import de.fraunhofer.aisec.codyze.analysis.passes.EdgeCachePass
import de.fraunhofer.aisec.codyze.analysis.passes.IdentifierPass
import de.fraunhofer.aisec.cpg.TranslationConfiguration
import de.fraunhofer.aisec.cpg.TranslationManager
import java.io.*
import java.lang.Exception
import java.util.ArrayList
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException
import java.util.function.Consumer
import kotlin.Throws
import org.junit.jupiter.api.*

internal class AnalysisServerBotanTest {
    @Test
    fun markModelTest() {
        val markModel = server!!.markModel
        Assertions.assertNotNull(markModel)
        val rules = markModel.rules
        Assertions.assertEquals(7, rules.size)
        val ents = markModel.entities
        Assertions.assertEquals(9, ents.size)
    }

    @Test
    fun markEvaluationTest() {
        val ctx = result
        Assertions.assertNotNull(ctx)
        val findings: MutableList<String> = ArrayList()
        Assertions.assertNotNull(ctx!!.findings)
        ctx.findings.forEach(Consumer { x: Finding -> findings.add(x.toString()) })
        println("Findings")
        for (finding in findings) {
            println(finding)
        }
    }

    companion object {
        private lateinit var server: AnalysisServer
        private var result: AnalysisContext? = null
        @BeforeAll
        @Throws(Exception::class)
        fun startup() {
            val classLoader = AnalysisServerBotanTest::class.java.classLoader
            var resource = classLoader.getResource("mark_cpp/symm_block_cipher.cpp")
            Assertions.assertNotNull(resource)
            val cppFile = File(resource.file)
            Assertions.assertNotNull(cppFile)
            resource = classLoader.getResource("mark/PoC_MS1/Botan_AutoSeededRNG.mark")
            Assertions.assertNotNull(resource)
            val markPoC1 = File(resource.file)
            Assertions.assertNotNull(markPoC1)
            val markModelFiles = markPoC1.parent

            // Start an analysis server
            server =
                AnalysisServer.builder()
                    .config(
                        ServerConfiguration.builder()
                            .launchConsole(false)
                            .launchLsp(false)
                            .markFiles(markModelFiles)
                            .build()
                    )
                    .build()
            server.start()

            // Start the analysis (BOTAN Symmetric Example by Oliver)
            val translationManager = newAnalysisRun(cppFile)
            val analyze = server.analyze(translationManager)
            try {
                result = analyze[5, TimeUnit.MINUTES]
            } catch (t: TimeoutException) {
                analyze.cancel(true)
                throw t
            }
        }

        @AfterAll
        fun teardown() {
            // Stop the analysis server
            server!!.stop()
        }

        /**
         * Helper method for initializing an Analysis Run.
         *
         * @param sourceLocations
         * @return
         */
        private fun newAnalysisRun(vararg sourceLocations: File): TranslationManager {
            return TranslationManager.builder()
                .config(
                    TranslationConfiguration.builder()
                        .debugParser(true)
                        .failOnError(false)
                        .defaultPasses()
                        .registerPass(IdentifierPass())
                        .registerPass(EdgeCachePass())
                        .defaultLanguages()
                        .sourceLocations(*sourceLocations)
                        .build()
                )
                .build()
        }
    }
}
